### PR TITLE
refactor(store): Make the Store async.

### LINKED ISSUE #64: Consider making cache methods async
This would allow for `Store` implementations not stored in memory, such as:

https://docs.hivedb.dev/#/advanced/lazy_box?id=lazy-box

### PR DESCRIPTION
**Pull request by GP4cK**
Hello!

Here's a PR to make the `Store` async as we discussed on Discord.

BREAKING CHANGE: All the read/writeQuery/Fragment are now async

Closes #64

I tried running the example pokemon app in Chrome and it seemed to work fine.
I didn't do anything regarding the race conditions that you mentioned though (like when the network is faster than the cache).
Do you think this could be handled in the Operation widget for example?

### CODE REVIEW

File: packages/ferry/test/typed_links/fetch_policy_typed_link_test.dart [hunk 1]
```diff
@@ -32,18 +30,9 @@ void main() {
...
-        ..human.height = 1.71,
-    );
-
     when(mockLink.request(any, any)).thenAnswer(
-      (_) => Stream.fromIterable([
-        Response(data: data.toJson(), response: {}),
-        Response(data: newData.toJson(), response: {}),
-      ]).interval(Duration(milliseconds: 10)),
+      (_) => Stream.value(Response(data: data.toJson(), response: {})).delay(
+          const Duration(milliseconds: 0)), //add delay to reliably test caching
```
Comment (GP4cK) [inline packages/ferry/test/typed_links/fetch_policy_typed_link_test.dart:35]: Here I reverted the code back to before https://github.com/gql-dart/ferry/pull/146.
It seems strange to me that you would get 2 responses from the server...

The tests to make sure we can return multiple responses from link are still there though

File: packages/ferry/test/typed_links/update_cache_typed_link_test.dart [hunk 2]
```diff
@@ -121,20 +118,24 @@ void main() {
...
+      expect(await cache.readQuery(reviewsReq), equals(null));
 
       requestController.add(createReviewReq);
       linkController
           .add(Response(data: createReviewData.toJson(), response: {}));
       await queue.next;
 
-      expect(cache.readQuery(reviewsReq)!.reviews!.length, equals(1));
+      // Wait for the store to finish writing.
+      await Future.delayed(Duration(milliseconds: 0));
```
Comment (GP4cK) [inline packages/ferry/test/typed_links/update_cache_typed_link_test.dart:129]: Without this line, the Store "doesn't have time" to write the data...
If you have a better approach, I'd be happy to hear about it.

**Comment by smkhalsa** [comment]
What's the use case driving this PR?

I'm interested to know how this would impact performance, especially with many consecutive reads and writes.

There may also be use cases that require sync access.

Maybe async methods could be added in addition to the sync ones?

**Comment by GP4cK** [comment]
> What's the use case driving this PR?

For me, it all started from this issue: https://github.com/gql-dart/ferry/issues/343.
TLDR: The sync implementation of Hive stores data in-memory which grows forever. I wanted to try 2 thing:
1. Use a LazyBox (but for that, I would need the cache to support async access
2. [Isar](https://github.com/isar/isar) which is recommended on the Hive page since the package is no longer actively maintained. Isar has both sync and async methods but the [sync methods are not supported on web](https://github.com/isar/isar/issues/380).

> I'm interested to know how this would impact performance, especially with many consecutive reads and writes.

I could do a benchmark for that. I could write a test which does X number of read / write and time it with both implementation. What do you think?

> Maybe async methods could be added in addition to the sync ones?

I don't think that would be possible with the way the cache/store are implemented.

**Comment by knaeckeKami** [comment]
Ok, so first: : I don't think that making the store async will to anything about https://github.com/gql-dart/ferry/issues/343 .
The issue seems to occur even with the MemoryStore which is basically just a BehaviourSubject of a Map.
So I don't see why switching from Hive to Isar would do anything.

That being said, if reading and writing from the Cache was async, we could hook other persistence implementations like Isar as store and even let the Cache run in another isolate which might improve performance for Flutter Apps where the Cache does too much work and causes frame drops currently.
So I am open to the idea in general.

But I see some challenges:

- Correctness: If reading and writing from the cache is asynchronous, we potentially open a huge can of worms. It would make many things much more complicated to get right. Things that need to be considered include: 
   - Race conditions (what if the network is quicker than the cache? we probably should not deliver outdated data from the cache AFTER the network data)
   - consistency: With an async Cache, there can be concurrent calls to `readQuery()` and `writeQuery()` (and their fragment counterparts). Without proper management of this, inconsistent state could be read from the cache. We probably want to avoid that. So there probably should be some locking mechanism (or Optimistic concurrency control)
 -  Breaking change: of course, this would be a large breaking change. We might consider letting the user choose between a sync or an async Cache. Maybe add an AsyncCache client additionally while still supporting the current behaviour?
 - We should coordinate with the `graphql` package and find a solution that also works for them. I would like to avoid going forward with something that breaks their code, this would likely lead so a fork of normalize and damage the gql ecosystem. I see them considering something similar though in https://github.com/zino-hofmann/graphql-flutter/issues/1196 https://github.com/zino-hofmann/graphql-flutter/issues/1151 . pinging @vincenzopalazzo and @budde377

**Comment by budde377** [comment]
I'm not sure how we should solve this problem, but I agree that making the API async will complicate things. Currently, I'm thinking of writing a store built on a two-layered cache to solve this:

* **Layer 1, in-memory**: The first layer is an in-memory cache layer that might be capped at some size. It can then evict the oldest records at will while supporting instantaneous cache readback.
* **Layer 2, persistent** The reason why we want persistent storage is so that we can deliver instant content across from app launches. We don't need to read this back instantly. So why don't we write entries from the in-memory cache through to persistent storage (disk for all I care)? We can then re-populate the in-memory store on launch and implement similar measures to keep storage down.

This should meet our requirements and would be possible with the existing API.

Regarding performance, i.e. https://github.com/zino-hofmann/graphql-flutter/issues/1196, what I'm considering here is to move the whole client into an Isolate. This'll remove network requests, cache access, serialization, and equality checks away from the main thread.

**Comment by GP4cK** [comment]
Hey, thanks for being reactive.
> I don't see why switching from Hive to Isar would do anything. But if you can measure any differences with your async implementation, correct me please.

@knaeckeKami the memory issues are both in the implementation of Hive (at least the SyncBox) and the MemoryStore. When providing an empty store, the issue goes away (see my comment in the linked issue). Supporting an async cache / store would give the possibility to try more implementation but since it seems we're not going to go down that road, I don't think it's worth spending time on testing. I guess the next step is to try using the cache implementation of the graphql package, as you said in the other issue and I'll see if I get the same jank issue that @budde377 mentioned.

@budde377 isolates are not supported on the web so you may end up with 2 different implementations. Have you already started working on that 2-layer cache?

**Comment by budde377** [comment]
> @budde377 isolates are not supported on the web so you may end up with 2 different implementations. Have you already started working on that 2-layer cache?

Haven't started yet, but I assume the implementation should be fairly straightforward. If you want to pick this up, you should feel free to do so!

**Comment by vincenzopalazzo** [comment]
Mh! in our issue https://github.com/zino-hofmann/graphql-flutter/issues/1196 the problem was and is related to some big operation of JSON encoding as this report pension

>When profiling with the current stable version I noticed significantly long time spent on the decode from dart:convert. With the new beta version 0.11.0-beta.5 this time seems significantly shorter.

In this case, the solution is to move the work to another thread, and you can't do anything else. However, in this particular case, the async does not work well because the JSON can consume the power of your thread and your async workflow is stuck to wait a better time.

However, the async solution is interesting stuff that can make the cache usage more `async` and we do not waste time waiting in serialization time. It is an interesting breaking change and If you would like to explore more and waste your time I can spend some time in this experiment.

**Comment by GP4cK** [comment]
I'll close this PR since it doesn't seem we have a consensus on whether the store should be async or not and even if we did, the implementation would be more complex than the one I suggested.
Plus, we found the root cause of the memory leak in https://github.com/gql-dart/ferry/pull/385
